How many late days are you using for this assignment? 

In [91]:
# Imports and pip installations (if needed)
%pip install pandas numpy scikit-learn tabulate
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


# Part 1: Load the dataset

In [92]:
# Load the given datasets
cate = pd.read_csv("./chronic_kidney_disease_categorical.csv")
num = pd.read_csv("./chronic_kidney_disease_numerical.csv")
display(num.rename(columns={
    "age": "Age",
    "bp": "Blood Pressure",
    "bgr": "Blood Glucose Random",
    "bu": "Blood Urea",
    "sc": "Serum Creatinine",
    "sod": "Sodium",
    "pot": "Potassium",
    "hemo": "Hemoglobin",
    "pcv": "Packed Cell Volume",
    "wbcc": "White Blood Cell Count",
    "rbcc": "Red Blood Cell Count"
}) )
display(cate.rename(columns = {
    "al": "Albumin",
    "su": "Sugar",
    "rbc": "Red Blood Cells",
    "pc": "Pus Cell",
    "pcc": "Pus Cell Clumps",
    "ba": "Bacteria",
    "htn": "Hypertension",
    "dm": "Diabetes Mellitus",
    "cad": "Coronary Artery Disease",
    "appet": "Appetite",
    "pe": "Pedal Edema",
    "ane": "Anemia",
    "Target": "Class"
}))


,unique_id,Age,Blood Pressure,Blood Glucose Random,Blood Urea,Serum Creatinine,Sodium,Potassium,Hemoglobin,Packed Cell Volume,White Blood Cell Count,Red Blood Cell Count
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


,unique_id,Albumin,Sugar,Red Blood Cells,Pus Cell,Pus Cell Clumps,Bacteria,Hypertension,Diabetes Mellitus,Coronary Artery Disease,Appetite,Pedal Edema,Anemia,Class
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The data set located at the provided link is a collection of both categorical and numerical values from patient records, detailing both patients with chronic kidney disease and without it. The only feature in this set is a categorical data point called target, denoting whether or not the given patient has chronic kidney disease and the features are all of their test results (medical values such as Blood Pressure and what not). All of the features in the numerical data set are numerical values (as the name may suggest) and all of those in the categorical data set, aside from the unique id, are categorical. Age, BP (blood pressure), and BU (blood urea) are all numerical values because they can be assigned to any number where as al (Albumin), su (sugar), and rbc (red blood cells) are categorical as they can only be assigned to a specific set of values. 

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [93]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numDup = num[num.duplicated()]
if(len(numDup) > 0):
    print("Numerical has ", len(numDup), " duplicate rows")
    display(numDup)
# Drop these duplicate rows
nDNum = num.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cateDup = cate[cate.duplicated()]
if(len(cateDup) > 0):
    print("Categorical has ", len(cateDup), " duplicate rows")
    display(cateDup)
nDCate = cate.drop_duplicates()


Numerical has  30  duplicate rows


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
107,872034,70.0,90.0,184.0,98.6,3.3,138.0,3.9,5.8,NaN,NaN,NaN
149,721270,58.0,70.0,102.0,48.0,1.2,139.0,4.3,15.0,40.0,8100.0,4.9
171,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
173,441170,79.0,80.0,111.0,44.0,1.2,146.0,3.6,16.3,40.0,8000.0,6.4
185,261137,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,5800.0,6.2
194,769512,38.0,60.0,91.0,36.0,0.7,135.0,3.7,14.0,46.0,9100.0,5.8
213,354223,64.0,70.0,97.0,27.0,0.7,145.0,4.8,13.8,49.0,6400.0,4.8
244,302906,59.0,100.0,NaN,96.0,6.4,NaN,NaN,6.6,NaN,NaN,NaN
245,431871,57.0,60.0,132.0,18.0,1.1,150.0,4.7,15.4,42.0,11000.0,4.5
248,687021,52.0,80.0,99.0,25.0,0.8,135.0,3.7,15.0,52.0,6300.0,5.3


Categorical has  30  duplicate rows


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
231,856244,0.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,poor,no,no,ckd
401,628271,2.0,1.0,abnormal,normal,notpresent,notpresent,yes,yes,no,poor,no,yes,ckd
402,158929,1.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,ckd
403,240975,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
404,441170,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
405,337834,4.0,0.0,normal,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
406,118191,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
407,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
408,302906,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,no,yes,ckd
409,938027,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, they cant be because we removed all of the duplicates yet the numerical table has 10 less entries than the categorical one. I think the rows thatre missing from numerical will be given the value NaN for the columns found in numerical and do nothing to those found in categorical. 

This is exemplified in the row i copied below

240430	NaN	NaN	NaN	NaN	NaN	NaN	NaN	NaN	NaN	NaN	NaN	0.0	0.0	normal	normal	NaN	NaN	no	no	no	good	no	no	notckd

In [94]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged = pd.merge(nDNum, nDCate, on = 'unique_id', how = 'right')
#Print the combined dataset
pd.reset_option("display.max_columns")

#Allow me to view the data in its entirety to check my above answer. 
pd.set_option("display.max_columns", None)
#pd.set_option("display.max_rows", None) 
#pd.reset_option("display.max_columns")
#pd.reset_option("display.max_rows")

display(merged)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,54.0,60.0,125.0,21.0,1.3,137.0,3.4,15.0,46.0,NaN,NaN,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,35.0,70.0,82.0,36.0,1.1,150.0,3.5,14.5,52.0,9400.0,6.1,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,65.0,70.0,139.0,29.0,1.0,NaN,NaN,10.5,32.0,NaN,NaN,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,5.0,NaN,NaN,16.0,0.7,138.0,3.2,8.1,NaN,NaN,NaN,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,42.0,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,789906,69.0,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
396,490253,43.0,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,492993,40.0,80.0,76.0,162.0,9.6,141.0,4.9,7.6,24.0,3800.0,2.8,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd
398,273038,47.0,60.0,117.0,22.0,1.2,138.0,3.5,13.0,45.0,5200.0,5.6,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [95]:
# Calculate the percentage of rows that contain atleast one missing value
missCount = merged.shape[0] - merged.dropna().shape[0]
missPercent = ((missCount/merged.shape[0])*100)
# Print %
print("The percentage of rows that have missing values is about", round(missPercent, 2), "%")
# Drop these rows from the dataset
merged.dropna(inplace= True)
# Print the Dataset
display(merged)

The percentage of rows that have missing values is about 61.75 %


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
4,881763,42.0,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
5,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,901528,73.0,80.0,118.0,44.0,0.7,137.0,3.5,14.8,45.0,9300.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,433613,50.0,80.0,97.0,40.0,0.6,150.0,4.5,14.2,48.0,10500.0,5.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,621332,59.0,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,poor,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
392,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
394,397388,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2,2.0,2.0,normal,normal,notpresent,present,yes,no,yes,good,no,no,ckd
395,789906,69.0,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
396,490253,43.0,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [96]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged.sort_values(by = ['Target'], inplace=True)
merged.reset_index(drop = True, inplace=True)

# Print the dataset
display(merged)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,534245,45.0,70.0,117.0,52.0,2.2,136.0,3.8,10.0,30.0,19100.0,3.7,2.0,0.0,normal,abnormal,present,notpresent,no,no,no,good,no,no,ckd
1,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
4,514721,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,530138,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,261137,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,5800.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,322882,33.0,80.0,128.0,38.0,0.6,135.0,3.9,13.1,45.0,6200.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [97]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target'] 
#I added al and su to the categorical columns because despite being categorical, they were not included in the encoding
merged = pd.get_dummies(merged, columns = categorical_columns, dtype= int)

# Print the dataset
display(merged)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,534245,45.0,70.0,117.0,52.0,2.2,136.0,3.8,10.0,30.0,19100.0,3.7,2.0,0.0,0,1,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0
1,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
3,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
4,514721,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4,2.0,0.0,1,0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,530138,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,261137,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,5800.0,6.2,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,322882,33.0,80.0,128.0,38.0,0.6,135.0,3.9,13.1,45.0,6200.0,4.5,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: In the particular example above, having a single binary var works great as all of the categorical values could only take one of two values in the first place. It actually may be better as it would use less memory and the other could be implied from one. If there was more than 2 possible values however, just having a single binary mapping would not work as there would be no way to indicate which of the 5 separate values the row holds with just a 1 or a 0.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [98]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    mean = merged[col].mean()
    std = merged[col].std()
    merged = merged[(merged[col] < mean + (3*std)) & (merged[col] > mean - (3*std))]

# Print the dataset
merged.reset_index(drop = True, inplace=True)
display(merged)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,514721,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4,2.0,0.0,1,0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0
2,124304,69.0,70.0,264.0,87.0,2.7,130.0,4.0,12.5,37.0,9600.0,4.1,3.0,4.0,0,1,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,1,0,1,0
3,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,7900.0,3.9,3.0,0.0,0,1,1,0,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
4,615697,6.0,60.0,94.0,67.0,1.0,135.0,4.9,9.9,30.0,16700.0,4.8,4.0,0.0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,489160,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
129,530138,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,261137,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,5800.0,6.2,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,322882,33.0,80.0,128.0,38.0,0.6,135.0,3.9,13.1,45.0,6200.0,4.5,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [99]:
# Normalize the all Numerical Attributes in the dataset.
scaler = StandardScaler()
merged[numerical_columns] = scaler.fit_transform(merged[numerical_columns])
# Print the dataset
display(merged)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,343710,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,0.519658,-2.301962,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,514721,0.494747,1.753327,0.170164,2.475789,3.577798,-1.663925,0.743327,-2.304078,-2.182119,-0.720859,-2.058397,2.0,0.0,1,0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0
2,124304,1.314898,-0.292221,3.165549,1.725293,0.881016,-1.839980,-0.516068,-0.824731,-1.024704,0.696875,-1.205921,3.0,4.0,0,1,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,1,0,1,0
3,980291,0.873279,-0.292221,0.014848,0.036678,0.206821,-0.783646,0.585903,-0.781221,-0.735350,-0.056296,-1.449485,3.0,0.0,0,1,1,0,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
4,615697,-2.659681,-1.314996,-0.606418,0.974798,-0.265116,-0.959702,0.900751,-1.955996,-2.037442,3.842472,-0.353444,4.0,0.0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,489160,-1.271733,-1.314996,-0.939238,0.336877,-0.602214,0.800855,1.058176,1.394289,1.000772,-0.676555,-0.109879,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
129,530138,-1.587175,0.730553,-0.229221,-0.263520,-0.197696,0.800855,-0.516068,-0.041547,-0.445996,-0.366426,-0.109879,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,261137,-0.830113,-1.314996,-0.362349,-0.075896,-0.602214,-0.959702,-0.673492,0.132493,-0.156643,-0.986684,1.351510,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,322882,-0.956290,0.730553,0.147976,-0.113421,-0.534794,-0.959702,-0.673492,-0.563670,0.132711,-0.809467,-0.718791,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: unique_id can be removed as we have already dropped all the duplicates so there is no reason for us to retain it. 

In [100]:
#Remove that column
merged.drop(columns = 'unique_id', inplace=True)
# Print the dataset
display(merged)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,0.519658,-2.301962,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,0.494747,1.753327,0.170164,2.475789,3.577798,-1.663925,0.743327,-2.304078,-2.182119,-0.720859,-2.058397,2.0,0.0,1,0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0
2,1.314898,-0.292221,3.165549,1.725293,0.881016,-1.839980,-0.516068,-0.824731,-1.024704,0.696875,-1.205921,3.0,4.0,0,1,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,1,0,1,0
3,0.873279,-0.292221,0.014848,0.036678,0.206821,-0.783646,0.585903,-0.781221,-0.735350,-0.056296,-1.449485,3.0,0.0,0,1,1,0,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
4,-2.659681,-1.314996,-0.606418,0.974798,-0.265116,-0.959702,0.900751,-1.955996,-2.037442,3.842472,-0.353444,4.0,0.0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,-1.271733,-1.314996,-0.939238,0.336877,-0.602214,0.800855,1.058176,1.394289,1.000772,-0.676555,-0.109879,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
129,-1.587175,0.730553,-0.229221,-0.263520,-0.197696,0.800855,-0.516068,-0.041547,-0.445996,-0.366426,-0.109879,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,-0.830113,-1.314996,-0.362349,-0.075896,-0.602214,-0.959702,-0.673492,0.132493,-0.156643,-0.986684,1.351510,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,-0.956290,0.730553,0.147976,-0.113421,-0.534794,-0.959702,-0.673492,-0.563670,0.132711,-0.809467,-0.718791,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [101]:
# Export the dataframe to a new csv file
merged.to_csv("cleaned_ckd_data_csv.csv")
# Export the dataframe to a new json file
merged.to_json("cleaned_ckd_data_json.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [102]:
display(merged.head(15))

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,0.519658,-2.301962,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,0.494747,1.753327,0.170164,2.475789,3.577798,-1.663925,0.743327,-2.304078,-2.182119,-0.720859,-2.058397,2.0,0.0,1,0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0
2,1.314898,-0.292221,3.165549,1.725293,0.881016,-1.839980,-0.516068,-0.824731,-1.024704,0.696875,-1.205921,3.0,4.0,0,1,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,1,0,1,0
3,0.873279,-0.292221,0.014848,0.036678,0.206821,-0.783646,0.585903,-0.781221,-0.735350,-0.056296,-1.449485,3.0,0.0,0,1,1,0,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
4,-2.659681,-1.314996,-0.606418,0.974798,-0.265116,-0.959702,0.900751,-1.955996,-2.037442,3.842472,-0.353444,4.0,0.0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0
5,0.431659,0.730553,2.056147,1.199946,1.690051,-0.607591,0.900751,-1.520894,-1.458735,-0.277817,-1.693050,3.0,1.0,0,1,1,0,0,1,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
6,-0.136138,-1.314996,0.924557,1.912917,1.285533,0.096632,-0.516068,-1.999507,-2.326796,2.912084,-2.301962,1.0,0.0,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
7,0.999456,2.776102,0.924557,0.486976,3.914895,-0.079424,0.428478,-2.826200,-2.616150,-0.233513,-2.058397,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
8,-0.009961,0.730553,-0.584229,4.577176,4.251993,-0.783646,-0.830916,-1.999507,-1.748088,-0.499338,-2.058397,4.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,0,1,1,0
9,1.314898,-0.292221,2.056147,2.063016,3.308119,-3.600537,-0.673492,-2.173547,-2.326796,1.538654,-2.180180,4.0,3.0,0,1,1,0,0,1,0,1,0,1,0,1,0,1,1,0,0,1,0,1,1,0


### Paste the markdown table here

Paste here: 
| age      | bp       | bgr      | bu       | sc       | sod      | pot       | hemo      | pcv      | wbcc     | rbcc     | al  | su  | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|----|----|--------------|------------|-------------|------------|---------------|------------|--------------|------------|--------|--------|------|------|------|------|-----------|-----------|------|------|------|------|-----------|--------------|
| 0.810190 | 0.730553 | 1.146438 | 4.014305 | 1.690051 | -0.959702 | 1.373024  | -2.913221 | -2.905503 | 0.519658  | -2.301962 | 2  | 0  | 1            | 0          | 1           | 0          | 1             | 0          | 1            | 0          | 0      | 1      | 0    | 1    | 0    | 1    | 0         | 1         | 0    | 1    | 0    | 1    | 1         | 0            |
| 0.494747 | 1.753327 | 0.170164 | 2.475789 | 3.577798 | -1.663925 | 0.743327  | -2.304078 | -2.182119 | -0.720859 | -2.058397 | 2  | 0  | 1            | 0          | 1           | 0          | 1             | 0          | 1            | 0          | 0      | 1      | 1    | 0    | 1    | 0    | 1         | 0         | 1    | 0    | 1    | 0    | 1         | 0            |
| 1.314898 | -0.292221 | 3.165549 | 1.725293 | 0.881016 | -1.839980 | -0.516068 | -0.824731 | -1.024704 | 0.696875  | -1.205921 | 3  | 4  | 0            | 1          | 1           | 0          | 1             | 0          | 1            | 0          | 0      | 1      | 0    | 1    | 0    | 1    | 1         | 0         | 0    | 1    | 1    | 0    | 1         | 0            |
| 0.873279 | -0.292221 | 0.014848 | 0.036678 | 0.206821 | -0.783646 | 0.585903  | -0.781221 | -0.735350 | -0.056296 | -1.449485 | 3  | 0  | 0            | 1          | 1           | 0          | 1             | 0          | 1            | 0          | 0      | 1      | 0    | 1    | 1    | 0    | 1         | 0         | 1    | 0    | 1    | 0    | 1         | 0            |
| -2.659681 | -1.314996 | -0.606418 | 0.974798 | -0.265116 | -0.959702 | 0.900751 | -1.955996 | -2.037442 | 3.842472  | -0.353444 | 4  | 0  | 1            | 0          | 1           | 0          | 1             | 0          | 0            | 1          | 1      | 0      | 1    | 0    | 1    | 0    | 1         | 0         | 0    | 1    | 1    | 0    | 1         | 0            |
| 0.431659 | 0.730553 | 2.056147 | 1.199946 | 1.690051 | -0.607591 | 0.900751  | -1.520894 | -1.458735 | -0.277817 | -1.693050 | 3  | 1  | 0            | 1          | 1           | 0          | 0             | 1          | 0            | 1          | 0      | 1      | 0    | 1    | 1    | 0    | 1         | 0         | 0    | 1    | 1    | 0    | 1         | 0            |
| -0.136138 | -1.314996 | 0.924557 | 1.912917 | 1.285533 | 0.096632  | -0.516068 | -1.999507 | -2.326796 | 2.912084  | -2.301962 | 1  | 0  | 0            | 1          | 0           | 1          | 1             | 0          | 1            | 0          | 0      | 1      | 0    | 1    | 1    | 0    | 1         | 0         | 1    | 0    | 1    | 0    | 1         | 0            |
| 0.999456 | 2.776102 | 0.924557 | 0.486976 | 3.914895 | -0.079424 | 0.428478  | -2.826200 | -2.616150 | -0.233513 | -2.058397 | 4  | 2  | 1            | 0          | 1           | 0          | 1             | 0          | 0            | 1          | 0      | 1      | 0    | 1    | 1    | 0    | 1         | 0         | 0    | 1    | 1    | 0    | 1         | 0            |
| -0.009961 | 0.730553 | -0.584229 | 4.577176 | 4.251993 | -0.783646 | -0.830916 | -1.999507 | -1.748088 | -0.499338 | -2.058397 | 4  | 0  | 0            | 1          | 1           | 0          | 1             | 0          | 1            | 0          | 0      | 1      | 1    | 0    | 1    | 0    | 1         | 0         | 1    | 0    | 1    | 0    | 0         | 1            |
| 1.314898 | -0.292221 | 2.056147 | 2.063016 | 3.308119 | -3.600537 | -0.673492 | -2.173547 | -2.326796 | 1.538654  | -2.180180 | 4  | 3  | 0            | 1          | 1           | 0          | 0             | 1          | 0            | 1          | 0      | 1      | 0    | 1    | 0    | 1    | 0         | 1         | 1    | 0    | 0    | 1    | 0         | 1            |
| 1.567253 | 2.776102 | 3.853379 | 1.837868 | 2.836183 | -0.079424 | -2.247736 | -2.260568 | -2.037442 | -0.455034 | -2.301962 | 3  | 2  | 1            | 0          | 1           | 0          | 0             | 1          | 1            | 0          | 0      | 1      | 0    | 1    | 0    | 1    | 0         | 1         | 0    | 1    | 1    | 0    | 1         | 0            |
| -1.713353 | 1.753327 | -0.317973 | -0.038371 | 0.206821 | -2.720258 | -1.303189 | -2.652160 | -3.050180 | 1.937392 | -1.449485 | 4  | 0  | 0            | 1          | 1           | 0          | 0             | 1          | 0            | 1          | 1      | 0      | 1    | 0    | 1    | 0    | 1         | 0         | 1    | 0    | 1    | 0    | 0         | 1            |
| 0.999456 | -1.314996 | 2.610848 | 0.637075 | 1.959729 | -0.607591 | 1.687873  | -2.130037 | -2.182119 | -0.233513 | -2.058397 | 4  | 1  | 1            | 0          | 1           | 0          | 1             | 0          | 0            | 1          | 0      | 1      | 0    | 1    | 1    | 0    | 0         | 1         | 0    | 1    | 1    | 0    | 1         | 0            |
| 0.494747 | -0.292221 | 1.967395 | -0.563718 | 0.206821 | -0.783646 | -0.830916 | 0.741636  | 1.145449  | 1.981696  | 0.620816  | 4  | 1  | 1            | 0          | 0           | 1          | 1             | 0          | 1            | 0          | 1      | 0      | 1    | 0    | 1    | 0    | 1         | 0         | 1    | 0    | 1    | 0    | 1         | 0            |
| -0.009961 | 0.730553 | 0.258916 | 3.676582 | 4.791349 | -1.487869 | 1.845297  | -1.781956 | -1.169381 | -0.809467 | -1.327703 | 4  | 0  | 1            | 0          | 1           | 0          | 1             | 0          | 0            | 1          | 1      | 0      | 0    | 1    | 1    | 0    | 1         | 0         | 0    | 1    | 1    | 0    | 1         | 0            |
| 1.567253 | 2.776102 | 3.853379 | 1.837868 | 2.836183 | -0.079424 | -2.247736 | -2.260568 | -2.037442 | -0.455034 | -2.301962 | 3  | 2  | 1            | 0          | 1           | 0          | 0             | 1          | 1            | 0          | 0      | 1      | 0    | 1    | 0    | 1    | 0         | 1         | 0    | 1    | 1    | 0    | 1         | 0            |
| -1.713353 | 1.753327 | -0.317973 | -0.038371 | 0.206821 | -2.720258 | -1.303189 | -2.652160 | -3.050180 | 1.937392 | -1.449485 | 4  | 0  | 0            | 1          | 1           | 0          | 0             | 1          | 0            | 1          | 1      | 0      | 1    | 0    | 1    | 0    | 1         | 0         | 1    | 0    | 1    | 0    | 0         | 1            |
| 0.999456 | -1.314996 | 2.610848 | 0.637075 | 1.959729 | -0.607591 | 1.687873  | -2.130037 | -2.182119 | -0.233513 | -2.058397 | 4  | 1  | 1            | 0          | 1           | 0          | 1             | 0          | 0            | 1          | 0      | 1      | 0    | 1    | 1    | 0    | 0         | 1         | 0    | 1    | 1    | 0    | 1         | 0            |
| 0.494747 | -0.292221 | 1.967395 | -0.563718 | 0.206821 | -0.783646 | -0.830916 | 0.741636  | 1.145449  | 1.981696  | 0.620816  | 4  | 1  | 1            | 0          | 0           | 1          | 1             | 0          | 1            | 0          | 1      | 0      | 1    | 0    | 1    | 0    | 1         | 0         | 1    | 0    | 1    | 0    | 1         | 0            |
| -0.009961 | 0.730553 | 0.258916 | 3.676582 | 4.791349 | -1.487869 | 1.845297  | -1.781956 | -1.169381 | -0.809467 | -1.327703 | 4  | 0  | 1            | 0          | 1           | 0          | 1             | 0          | 0            | 1          | 1      | 0      | 0    | 1    | 1    | 0    | 1         | 0         | 0    | 1    | 1    | 0    | 1         | 0            |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

I personally had it go 5 lines at a time. 

Note I have no idea why its not showing up as readable in the pdf, it works in my vscode and im actually slamming my head into the wall with gpt over this one. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [ ]:
# Code to rename all the columns in the dataset
# Dictionary to map abbreviated names to full names
abbreviation_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'al': 'albumin',
    'su': 'sugar',
    'rbc_abnormal': 'red blood cells abnormal',
    'rbc_normal': 'red blood cells normal',
    'pc_abnormal': 'pus cell abnormal',
    'pc_normal': 'pus cell normal',
    'pcc_notpresent': 'pus cell clumps not present',
    'pcc_present': 'pus cell clumps present',
    'ba_notpresent': 'bacteria not present',
    'ba_present': 'bacteria present',
    'htn_no': 'hypertension no',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'target chronic kidney disease',
    'Target_notckd': 'target not chronic kidney disease'
}

#note, i told gpt about the encoding

# Rename columns using the mapping
merged.rename(columns=abbreviation_mapping, inplace=True)
display(merged)


,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,white blood cell count,red blood cell count,albumin,sugar,red blood cells abnormal,red blood cells normal,pus cell abnormal,pus cell normal,pus cell clumps not present,pus cell clumps present,bacteria not present,bacteria present,hypertension no,hypertension yes,diabetes mellitus no,diabetes mellitus yes,coronary artery disease no,coronary artery disease yes,appetite good,appetite poor,pedal edema no,pedal edema yes,anemia no,anemia yes,target chronic kidney disease,target not chronic kidney disease
0,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,0.519658,-2.301962,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,0.494747,1.753327,0.170164,2.475789,3.577798,-1.663925,0.743327,-2.304078,-2.182119,-0.720859,-2.058397,2.0,0.0,1,0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0
2,1.314898,-0.292221,3.165549,1.725293,0.881016,-1.839980,-0.516068,-0.824731,-1.024704,0.696875,-1.205921,3.0,4.0,0,1,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,1,0,1,0
3,0.873279,-0.292221,0.014848,0.036678,0.206821,-0.783646,0.585903,-0.781221,-0.735350,-0.056296,-1.449485,3.0,0.0,0,1,1,0,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
4,-2.659681,-1.314996,-0.606418,0.974798,-0.265116,-0.959702,0.900751,-1.955996,-2.037442,3.842472,-0.353444,4.0,0.0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,-1.271733,-1.314996,-0.939238,0.336877,-0.602214,0.800855,1.058176,1.394289,1.000772,-0.676555,-0.109879,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
129,-1.587175,0.730553,-0.229221,-0.263520,-0.197696,0.800855,-0.516068,-0.041547,-0.445996,-0.366426,-0.109879,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,-0.830113,-1.314996,-0.362349,-0.075896,-0.602214,-0.959702,-0.673492,0.132493,-0.156643,-0.986684,1.351510,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,-0.956290,0.730553,0.147976,-0.113421,-0.534794,-0.959702,-0.673492,-0.563670,0.132711,-0.809467,-0.718791,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
    ```sql
    SELECT Target, COUNT(*) AS count
    FROM your_table_name
    GROUP BY Target;
    ```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [90]:
result = merged.groupby('target chronic kidney disease').size().reset_index(name='count')

# Display the result
print(result)

   target chronic kidney disease  count
0                              0    111
1                              1     22


*Describe what the above code does here*

takes the dataframe, groups it by identical values found in the given column, returning a series, takes the size of eaach series and puts their value into a column called count. 